pkg/conversion/generator.go

```diff
@@ -0,0 +1,414 @@
+	// inside appropriate conversion functions.
+	return nil
+}
+
+func (g *generator) generateConversionsForStruct(inType, outType reflect.Type) error {
+	for i := 0; i < inType.NumField(); i++ {
+		inField := inType.Field(i)
+		outField, found := outType.FieldByName(inField.Name)
+		if !found {
+			return fmt.Errorf("couldn't find a corresponding field %v in %v", inField.Name, outType)
+		}
+		if inField.Type.Kind() != outField.Type.Kind() {
+			return fmt.Errorf("cannot convert types of different kinds: %v %v", inField, outField)
+		}
+		if isComplexType(inField.Type) {
+			if err := g.generateConversionsBetween(inField.Type, outField.Type); err != nil {
+				return err
+			}
+		}
+	}
+	g.convertibles[inType] = outType
+	return nil
+}
+
+func (g *generator) WriteConversionFunctions(w io.Writer) error {
+	indent := 2
+	for inType, outType := range g.convertibles {
+		// All types in g.convertibles are structs.
+		if inType.Kind() != reflect.Struct {
+			return fmt.Errorf("non-struct conversions are not-supported")
+		}
+		if err := g.writeConversionForStruct(w, inType, outType, indent); err != nil {
+			return err
+		}
+		if err := g.writeConversionForStruct(w, outType, inType, indent); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (g *generator) typeName(inType reflect.Type) string {
+	switch inType.Kind() {
+	case reflect.Map:
+		return fmt.Sprintf("map[%s]%s", g.typeName(inType.Key()), g.typeName(inType.Elem()))
+	case reflect.Slice:
+		return fmt.Sprintf("[]%s", g.typeName(inType.Elem()))
+	default:
+		typeWithPkg := fmt.Sprintf("%s", inType)
+		slices := strings.Split(typeWithPkg, ".")
+		if len(slices) == 1 {
+			// Default package.
+			return slices[0]
+		}
+		if len(slices) == 2 {
+			pkg := slices[0]
+			if val, found := g.pkgOverwrites[pkg]; found {
+				pkg = val
+			}
+			if pkg != "" {
+				pkg = pkg + "."
+			}
+			return pkg + slices[1]
+		}
+		panic("Incorrect type name: " + typeWithPkg)
+	}
+}
+
+func indentLine(w io.Writer, indent int) error {
+	indentation := ""
+	for i := 0; i < indent; i++ {
+		indentation = indentation + "\t"
+	}
+	_, err := io.WriteString(w, indentation)
+	return err
+}
+
+func writeLine(w io.Writer, indent int, line string) error {
+	if err := indentLine(w, indent); err != nil {
+		return err
+	}
+	if _, err := io.WriteString(w, line); err != nil {
+		return err
+	}
+	return nil
+}
+
+func writeHeader(w io.Writer, inType, outType string, indent int) error {
+	format := "func(in *%s, out *%s, s conversion.Scope) error {\n"
+	stmt := fmt.Sprintf(format, inType, outType)
+	return writeLine(w, indent, stmt)
+}
+
+func writeFooter(w io.Writer, indent int) error {
+	if err := writeLine(w, indent+1, "return nil\n"); err != nil {
+		return err
+	}
+	if err := writeLine(w, indent, "},\n"); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (g *generator) writeConversionForMap(w io.Writer, inField, outField reflect.StructField, indent int) error {
```

Comment:
Suggestion: look into the template library... I expect these would be much easier to understand as templates than this way, although I could be wrong.